ovos_utils/gui.py

```diff
@@ -9,6 +13,347 @@ def is_gui_installed():
+
+    # GUI event handlers
+    # user can/should subclass this
+    def on_idle(self, namespace):
+        pass
+
+    def on_active(self, namespace):
+        pass
+
+    def on_new_page(self, page, namespace, index):
+        pass
+
+    def on_gui_value(self, namespace, key, value):
+        pass
+
+    ######################################################################
+    # GUI client API
+    # TODO see how much of this can be removed
+    @staticmethod
+    def _get_page_data(message):
+        """ Extract page related data from a message.
+
+        Args:
+            message: messagebus message object
+        Returns:
+            tuple (page, namespace, index)
+        Raises:
+            ValueError if value is missing.
+        """
+        data = message.data
+        # Note:  'page' can be either a string or a list of strings
+        if 'page' not in data:
+            raise ValueError("Page missing in data")
+        if 'index' in data:
+            index = data['index']
+        else:
+            index = 0
+        page = data.get("page", "")
+        namespace = data.get("__from", "")
+        return page, namespace, index
+
+    def _set(self, namespace, name, value):
+        """ Perform the send of the values to the connected GUIs. """
+        if namespace not in self._datastore:
+            self._datastore[namespace] = {}
+        if self._datastore[namespace].get(name) != value:
+            self._datastore[namespace][name] = value
+
+    def __find_namespace(self, namespace):
+        for i, skill in enumerate(self._loaded):
+            if skill[0] == namespace:
+                return i
+        return None
+
+    def __insert_pages(self, namespace, pages):
+        """ Insert pages into the namespace
+
+        Args:
+            namespace (str): Namespace to add to
+            pages (list):    Pages (str) to insert
+        """
+        LOG.debug("Inserting new pages")
+        # Insert the pages into local reprensentation as well.
+        updated = self.Namespace(self._loaded[0].name,
+                                 self._loaded[0].pages + pages)
+        self._loaded[0] = updated
+
+    def __remove_page(self, namespace, pos):
+        """ Delete page.
+
+        Args:
+            namespace (str): Namespace to remove from
+            pos (int):      Page position to remove
+        """
+        LOG.debug("Deleting {} from {}".format(pos, namespace))
+        # Remove the page from the local reprensentation as well.
+        self._loaded[0].pages.pop(pos)
+
+    def __insert_new_namespace(self, namespace, pages):
+        """ Insert new namespace and pages.
+
+        This first sends a message adding a new namespace at the
+        highest priority (position 0 in the namespace stack)
+
+        Args:
+            namespace (str):  The skill namespace to create
+            pages (str):      Pages to insert (name matches QML)
+        """
+        if time.time() - self.idle_ts > 1:
```

Comment:
tests in my laptop showed this value to be around 0.2 seconds, usually less.

1 second is to provide a safety margin, but i would like to further reduce this value after testing in other devices, just leaving this note here to revisit after mark2 is launched

i also expect the internal api will change and be less mark2 specific once enclosures are sorted out, would be much cleaner if we could add a message to upstream since this check is very hacky